python/tests/graph/test_input_graph.py

```diff
@@ -0,0 +1,223 @@
+
+                colname = '{}_replica{:03d}'.format(self.output_col, i)
+                tnsr_op_name = '{}_replica{:03d}'.format(self.output_op_name, i)
+                self.output_mapping[tnsr_op_name] = colname
+                self.fetch_names.append(tnsr_op_name + ':0')
+        else:
+            self.input_mapping = {self.input_col: self.input_op_name}
+            self.feed_names = [self.input_op_name + ':0']
+            self.output_mapping = {self.output_op_name: self.output_col}
+            self.fetch_names = [self.output_op_name + ':0']
+
+    @contextmanager
+    def _run_test_in_tf_session(self):
+        """ [THIS IS NOT A TEST]: encapsulate general test workflow """
+
+        # Build the TensorFlow graph
+        graph = tf.Graph()
+        with tf.Session(graph=graph) as sess, graph.as_default():
+            # Build test graph and transformers from here
+            yield sess
+
+            ref_feed = tfx.get_tensor(graph, self.input_op_name)
+            ref_fetch = tfx.get_tensor(graph, self.output_op_name)
+
+            def check_input_graph(tgt_gdef, test_idx):
+                namespace = 'TEST_TGT_NS{:03d}'.format(test_idx)
+                tf.import_graph_def(tgt_gdef, name=namespace)
+                tgt_feed = tfx.get_tensor(graph, '{}/{}'.format(namespace, self.input_op_name))
+                tgt_fetch = tfx.get_tensor(graph, '{}/{}'.format(namespace, self.output_op_name))
+
+                for _ in range(10):
+                    local_data = np.random.randn(31, self.vec_size)
+                    ref_out = sess.run(ref_fetch, feed_dict={ref_feed: local_data})
+                    tgt_out = sess.run(tgt_fetch, feed_dict={tgt_feed: local_data})
+                    self.assertTrue(np.allclose(ref_out, tgt_out))
+
+            for test_idx, input_graph in enumerate(self.input_graphs):
+                check_input_graph(input_graph.graph_def, test_idx)
+
+
+    def test_build_from_tf_graph(self):
+        """ Build TFTransformer from tf.Graph """
+        with self._run_test_in_tf_session() as sess:
+            # Begin building graph
+            x = tf.placeholder(tf.float64, shape=[None, self.vec_size], name=self.input_op_name)
+            _ = tf.reduce_mean(x, axis=1, name=self.output_op_name)
+
+            gin = TFInputGraph.fromGraph(sess.graph, sess, self.feed_names, self.fetch_names)
+            self.input_graphs.append(gin)
+            # End building graph
+
+    def test_build_from_saved_model(self):
+        """ Build TFTransformer from saved model """
+        # Setup saved model export directory
+        saved_model_root = self.model_output_root
+        saved_model_dir = os.path.join(saved_model_root, 'saved_model')
+        serving_tag = "serving_tag"
+        serving_sigdef_key = 'prediction_signature'
+        builder = tf.saved_model.builder.SavedModelBuilder(saved_model_dir)
+
+        with self._run_test_in_tf_session() as sess:
+            # Model definition: begin
```

Comment:
can all the tests here use the same graph built by calling a helper function? e.g.
```
def _build_graph():
  g = tf.Graph()
  with g.as_default():
    x = tf.placeholder(tf.float64, shape=[None, self.vec_size], name=self.input_op_name)
    w = tf.Variable(tf.random_normal([self.vec_size], dtype=tf.float64),
                           dtype=tf.float64, name='varW')
    z = tf.reduce_mean(x * w, axis=1, name=self.output_op_name)
  return g

...

  def test_build_...():
    graph = _build_graph();
    with tf.Session(graph=graph) as sess:
      ...
```

Reply:
Well, this is essentially what we are doing here, isn't it? 
Testing functions also serve as example/documentation. Here having the whole workflow in the same place makes it easy for users to understand how to use our library.